cornac/models/nmf/recom_nmf.pyx

```diff
@@ -0,0 +1,291 @@
+            int num_threads = self.num_threads
+            floating lr = self.learning_rate
+
+            floating lambda_u = self.lambda_u
+            floating lambda_v = self.lambda_v
+            floating lambda_bu = self.lambda_bu
+            floating lambda_bi = self.lambda_bi
+            floating mu = self.global_mean
+
+            bool use_bias = self.use_bias
+            bool verbose = self.verbose
+
+            floating[:, :] U_numerator = np.zeros((num_users, num_factors), dtype=np.float32)
+            floating[:, :] V_numerator = np.zeros((num_items, num_factors), dtype=np.float32)
+            floating[:, :] U_denominator = np.zeros((num_users, num_factors), dtype=np.float32)
+            floating[:, :] V_denominator = np.zeros((num_items, num_factors), dtype=np.float32)
+
+            floating loss, r, r_pred, error, eps = 1e-9
+
+            long u, i, f, j
+            long n_ratings
+
+        from tqdm import trange
+        progress = trange(max_iter, disable=not verbose)
+        for epoch in progress:
+
+            for j in prange(num_ratings, nogil=True, num_threads=num_threads):
+                u, i, r = rid[j], cid[j], val[j]
+
+                # predict rating
+                r_pred = mu + Bu[u] + Bi[i]
+                for f in range(num_factors):
+                    r_pred = r_pred + U[u, f] * V[i, f]
+
+                error = r - r_pred
+
+                # update biases
+                if use_bias:
+                    Bu[u] += lr * (error - lambda_bu * Bu[u])
+                    Bi[i] += lr * (error - lambda_bi * Bi[i])
+
+                # compute numerators and denominators
+                for f in range(num_factors):
+                    U_numerator[u, f] += r * V[i, f]
+                    U_denominator[u, f] += r_pred * V[i, f]
+                    V_numerator[i, f] += r * U[u, f]
+                    V_denominator[i, f] += r_pred * U[u, f]
+
+            # update user factors
+            for u in range(num_users):
+                n_ratings = user_counts[u]
+                for f in range(num_factors):
+                    U_denominator[u, f] += n_ratings * lambda_u * U[u, f] + eps
+                    U[u, f] *= U_numerator[u, f] / U_denominator[u, f]
+
+            # update item factors
+            for i in range(num_items):
+                n_ratings = item_counts[i]
+                for f in range(num_factors):
+                    V_denominator[i, f] += n_ratings * lambda_v * V[i, f] + eps
+                    V[i, f] *= V_numerator[i, f] / V_denominator[i, f]
+
+            # reset all numerators and denominators
+            U_numerator = np.zeros((num_users, num_factors), dtype=np.float32)
+            V_numerator = np.zeros((num_items, num_factors), dtype=np.float32)
+            U_denominator = np.zeros((num_users, num_factors), dtype=np.float32)
+            V_denominator = np.zeros((num_items, num_factors), dtype=np.float32)
+
+            loss = 0.
```

Comment:
Why do we need to calculate the loss separately?
Why don't we merge it with the previous loops?

Reply:
We cannot calculate loss within prange loop